core/utils/copy.h

```diff
@@ -0,0 +1,204 @@
+    case 2:
+      memcpy(d, s, 2);
+      break;
+    case 1:
+      memcpy(d, s, 1);
+      break;
+  }
+}
+
+// Copies "bytes" data from "src" to "dst".
+// Same as memcpy() and rte_memcpy(), but significantly faster for both
+// aligned/unaligned buffers. Performs best if aligned, of course.
+// bytes can be 0.
+//
+// NOTE: When "sloppy" is set, it may copy more than "bytes", up to additional
+// 31 bytes. It will generate much smaller and usually faster code. Use this
+// option only if overwriting some data at the end is acceptable, such as
+// rewriting the payload data of class Packet.
+static inline void Copy(void *__restrict__ dst, const void *__restrict__ src,
+                        size_t bytes, bool sloppy = false) {
+#if __AVX2__
+  using block_t = __m256i;
+  auto copy_block = [](void *__restrict__ d, const void *__restrict__ s) {
+    Copy32(d, s);
+  };
+#else
+  using block_t = __m128i;
+  auto copy_block = [](void *__restrict__ d, const void *__restrict__ s) {
+    Copy16(d, s);
+  };
+#endif
+
+  const size_t block_size = sizeof(block_t);
+  uintptr_t dst_u = reinterpret_cast<uintptr_t>(dst);
+  uintptr_t src_u = reinterpret_cast<uintptr_t>(src);
+
+  if (bytes <= 64 && !sloppy) {
+    CopySmall(dst, src, bytes);
+    return;
+  }
+
+  // Align dst on a cache line if buffer is big yet misaligned.
+  if (bytes >= 256 && (dst_u % block_size) != 0) {
+    // Copy "block_t" bytes, but proceed with only "offset" bytes.
+    copy_block(reinterpret_cast<block_t *__restrict__>(dst),
+               reinterpret_cast<const block_t *__restrict__>(src));
+
+    uintptr_t offset = block_size - (dst_u % block_size);
+    dst = reinterpret_cast<decltype(dst)>(dst_u + offset);
+    src = reinterpret_cast<decltype(src)>(src_u + offset);
+    bytes -= offset;
+  }
+
+  auto *d = reinterpret_cast<block_t *__restrict__>(dst);
+  auto *s = reinterpret_cast<const block_t *__restrict__>(src);
+
+  size_t num_blocks = (sloppy ? bytes + block_size - 1 : bytes) / block_size;
+  size_t num_loops = num_blocks / 8;
+
+  while (num_loops--) {
+    copy_block(d + 0, s + 0);
+    copy_block(d + 1, s + 1);
+    copy_block(d + 2, s + 2);
+    copy_block(d + 3, s + 3);
+    copy_block(d + 4, s + 4);
+    copy_block(d + 5, s + 5);
+    copy_block(d + 6, s + 6);
+    copy_block(d + 7, s + 7);
+    d += 8;
+    s += 8;
+  }
+
+  // Copy the leftover. No block to copy if remainder is 0
+  size_t leftover_blocks = num_blocks % 8;
+
+  switch (leftover_blocks) {
+    case 7:
+      copy_block(d + 6, s + 6);
+    case 6:
+      copy_block(d + 5, s + 5);
+    case 5:
+      copy_block(d + 4, s + 4);
+    case 4:
+      copy_block(d + 3, s + 3);
+    case 3:
+      copy_block(d + 2, s + 2);
+    case 2:
+      copy_block(d + 1, s + 1);
+    case 1:
+      copy_block(d + 0, s + 0);
+  }
+
+  if (!sloppy && (bytes % block_size) != 0) {
+    size_t fringe = bytes % block_size;
+    dst_u = reinterpret_cast<decltype(dst_u)>(d + leftover_blocks);
```

Comment:
Not a big deal but the use of `decltype(dst_u)` worries me here, since it is unambiguously `uintptr_t`. Is there a reason to use it here. I kind of understand for line 196 (it is much longer), but it seems strange here.